File: js/bucket_sort.js
// only works with non-negative numbers
function bucketSort(array) {
    let n = Number.MIN_VALUE;
    for (let i = 0; i < array.length; i++) {
        if (array[i] > n) {
            n = array[i];
        }
    }

    const buckets = new Array(n + 1).fill(null);

    for (let i = 0; i < array.length; i++) {
        let index = array[i];
        if (buckets[index] == null) {
            buckets[index] = [];
        }

        buckets[index].push(array[i]);
    }

    let k = 0;
    for (let i = 0; i < buckets.length; i++) {
        const bucket = buckets[i];
        if (bucket) {
            for (let j = 0; j < bucket.length; j++) {
                array[k++] = bucket[j];
            }
        }
    }
}

// works with negative numbers
function bucketSort2(array) {
    let min = Number.MAX_VALUE;
    let max = Number.MIN_VALUE;

    for (let i = 0; i < array.length; i++) {
        if (array[i] < min) {
            min = array[i];
        }

        if (array[i] > max) {
            max = array[i];
        }
    }

    const range = max - min + 1;
    const offset = 0 - min;

    const buckets = new Array(range).fill(null);

    for (let i = 0; i < array.length; i++) {
        const key = array[i] + offset;
        if (buckets[key] == null) {
            buckets[key] = [];
        }
        buckets[key].push(array[i]);
    }

    let k = 0;
    for (let i = 0; i < buckets.length; i++) {
        if (buckets[i] != null) {
            const bucket = buckets[i];
            for (let j = 0; j < bucket.length; j++) {
                array[k++] = bucket[j];
            }
        }
    }
}

// without array
function bucketSort3(list) {
    let min = (max = list[0]);
    for (let i = 1; i < list.length; i++) {
        if (list[i] < min) {
            min = list[i];
        }
        if (list[i] > max) {
            max = list[i];
        }
    }

    const range = max - min + 1;
    const offset = 0 - min;

    const buckets = new Array(range).fill(0);

    for (let i = 0; i < list.length; i++) {
        const key = list[i] + offset;
        buckets[key] = buckets[key] + 1;
    }

    let k = 0;
    for (let i = 0; i < buckets.length; i++) {
        if (buckets[i] > 0) {
            for (let j = 0; j < buckets[i]; j++) {
                list[k++] = i - offset;
            }
        }
    }
}

function testBucketSort() {
    output.innerHTML += "<h3>testing bucket sort in javascript...</h3>";
    const unsortedArray = getRandArray(10, 1);
    output.innerHTML += `<p>unsorted array -> [${unsortedArray}]</p>`;
    bucketSort(unsortedArray);
    output.innerHTML += `<p>sorted array -> [${unsortedArray}]</p>`;
}

function testBucketSort2() {
    output.innerHTML += "<h3>testing bucket sort 2 in javascript...</h3>";
    const unsortedArray = getRandArray(9, -2);
    output.innerHTML += `<p>unsorted array -> [${unsortedArray}]</p>`;
    bucketSort2(unsortedArray);
    output.innerHTML += `<p>sorted array -> [${unsortedArray}]</p>`;
}

function testBucketSort3() {
    output.innerHTML += "<h3>testing bucket sort 3 in javascript...</h3>";
    const unsortedArray = getRandArray(9, -2);
    output.innerHTML += `<p>unsorted array -> [${unsortedArray}]</p>`;
    bucketSort3(unsortedArray);
    output.innerHTML += `<p>sorted array -> [${unsortedArray}]</p>`;
}

testBucketSort();
testBucketSort2();
testBucketSort3();

File: python/bucket_sort.py
# doesn't work with negative numbers
def bucket_sort(list):
    n = 0
    for val in list:
        if val > n:
            n = val
    
    buckets = [None] * (n + 1)
    
    for val in list:
        if buckets[val] is None:
            buckets[val] = []
        
        buckets[val].append(val)

    k = 0
    for bucket in buckets:
        if bucket is not None:
            for val in bucket:
                list[k] = val
                k += 1


#works with negative numbers
def bucket_sort_2(list):
    min = max = list[0]

    for val in list:
        if val < min:
            min = val
        if val > max:
            max = val

    range = max - min + 1
    offset = 0 - min

    buckets = [None] * range

    for val in list:
        key = val + offset
        if buckets[key] == None:
            buckets[key] = []
        buckets[key].append(val)

    k = 0
    for bucket in buckets:
        if bucket is not None:
            for val in bucket:
                list[k] = val
                k += 1

# bucket sort without array list
def bucket_sort_3(list):
    min = list[0]
    max = list[0]
    for i in range(1, len(list)):
        if list[i] < min:
            min = list[i]
        if list[i] > max:
            max = list[i]
    
    r = max - min + 1
    offset = 0 - min

    buckets = [0] * r

    for val in list:
        key = val + offset
        buckets[key] = buckets[key] + 1
    
    k = 0
    for i in range(len(buckets)):
        if buckets[i] > 0:
            for _ in range(buckets[i]):
                list[k] = i - offset
                k += 1

print("testing bucket sort in python...")

unsortedArray = [4, 1, 5, 2, 8, 3, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
bucket_sort(unsortedArray)
print("sorted array -> " + str(unsortedArray))

print("testing bucket sort 2 in python...")

unsortedArray = [4, 1, -2, 5, 2, 0, 8, 3, -1, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
bucket_sort_2(unsortedArray)
print("sorted array -> " + str(unsortedArray))

print("testing bucket sort 3 in python...")

unsortedArray = [4, 1, -2, 5, 2, 0, 8, 3, -1, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
bucket_sort_3(unsortedArray)
print("sorted array -> " + str(unsortedArray))